protocols/autonat/src/v2/server/behaviour.rs

```diff
@@ -0,0 +1,165 @@
+use std::{
+    collections::{HashMap, VecDeque},
+    io,
+    task::{Context, Poll},
+};
+
+use crate::v2::server::handler::dial_request::DialBackStatus;
+use either::Either;
+use libp2p_core::{transport::PortUse, Endpoint, Multiaddr};
+use libp2p_identity::PeerId;
+use libp2p_swarm::{
+    dial_opts::DialOpts, ConnectionDenied, ConnectionHandler, ConnectionId, DialFailure, FromSwarm,
+    NetworkBehaviour, ToSwarm,
+};
+use libp2p_swarm::{dial_opts::PeerCondition, ConnectionClosed};
+use rand_core::{OsRng, RngCore};
+
+use crate::v2::server::handler::{
+    dial_back,
+    dial_request::{self, DialBackCommand},
+    Handler,
+};
+
+pub struct Behaviour<R = OsRng>
+where
+    R: Clone + Send + RngCore + 'static,
+{
+    dialing_dial_back: HashMap<ConnectionId, DialBackCommand>,
+    pending_events: VecDeque<
+        ToSwarm<
+            <Self as NetworkBehaviour>::ToSwarm,
+            <<Self as NetworkBehaviour>::ConnectionHandler as ConnectionHandler>::FromBehaviour,
+        >,
+    >,
+    rng: R,
+}
+
+impl Default for Behaviour<OsRng> {
+    fn default() -> Self {
+        Self::new(OsRng)
+    }
+}
+
+impl<R> Behaviour<R>
+where
+    R: RngCore + Send + Clone + 'static,
+{
+    pub fn new(rng: R) -> Self {
+        Self {
+            dialing_dial_back: HashMap::new(),
+            pending_events: VecDeque::new(),
+            rng,
+        }
+    }
+}
+
+impl<R> NetworkBehaviour for Behaviour<R>
+where
+    R: RngCore + Send + Clone + 'static,
+{
+    type ConnectionHandler = Handler<R>;
+
+    type ToSwarm = StatusUpdate;
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        _connection_id: ConnectionId,
+        peer: PeerId,
+        _local_addr: &Multiaddr,
+        remote_addr: &Multiaddr,
+    ) -> Result<<Self as NetworkBehaviour>::ConnectionHandler, ConnectionDenied> {
+        Ok(Either::Right(dial_request::Handler::new(
+            peer,
+            remote_addr.clone(),
+            self.rng.clone(),
+        )))
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        connection_id: ConnectionId,
+        _peer: PeerId,
+        _addr: &Multiaddr,
+        _role_override: Endpoint,
+        _port_use: PortUse,
+    ) -> Result<<Self as NetworkBehaviour>::ConnectionHandler, ConnectionDenied> {
+        Ok(
+            if let Some(cmd) = self.dialing_dial_back.remove(&connection_id) {
+                Either::Left(dial_back::Handler::new(cmd))
+            } else {
+                Either::Left(dial_back::Handler::empty())
+            },
```

Comment:
I think this would be cleaner as a `match`.

Reply:
Refactored it.